File: cmd/basic-vad-web/static/js/dsp/fbank.js
// Port of fbank/fbank.go, fbank/mel.go, fbank/window.go.
//
// Kaldi-compatible log-Mel-filterbank features for the FSMN-VAD and
// FireRedVAD backends. Match `compute-fbank-feats` defaults: HTK mel scale,
// Povey or Hamming window, snip_edges=true, preemph=0.97, RemoveDCOffset.
//
// Float32-throughput pipeline (doubles internally where parity matters).

import { fftRadix2, nextPow2 } from './fft.js';

export const WINDOW_HAMMING = 'hamming';
export const WINDOW_POVEY = 'povey';

// Match fbank.Defaults() in Go.
export function fbankDefaults() {
    return {
        sampleRate: 16000,
        frameLengthMs: 25,
        frameShiftMs: 10,
        numMelBins: 80,
        windowType: WINDOW_POVEY,
        lowFreqHz: 20,
        highFreqHz: 0,        // 0 ⇒ sampleRate/2
        preemphCoeff: 0.97,
        removeDCOffset: true,
        snipEdges: true,
        energyFloor: 0,
        useLog: true,
    };
}

// makeWindow returns a Float64Array of length L.
function makeWindow(length, type) {
    const w = new Float64Array(length);
    if (length === 1) { w[0] = 1; return w; }
    const denom = length - 1;
    if (type === WINDOW_HAMMING) {
        for (let i = 0; i < length; i++) {
            w[i] = 0.54 - 0.46 * Math.cos(2 * Math.PI * i / denom);
        }
    } else {
        // Povey (default if unknown).
        for (let i = 0; i < length; i++) {
            const h = 0.5 - 0.5 * Math.cos(2 * Math.PI * i / denom);
            w[i] = Math.pow(h, 0.85);
        }
    }
    return w;
}

// HTK mel scale, used by Kaldi.
function hzToMel(hz) { return 1127.0 * Math.log(1.0 + hz / 700.0); }
function melToHz(mel) { return 700.0 * (Math.exp(mel / 1127.0) - 1.0); }

// makeMelFilterbank returns [numBins][fftBins] (array of Float64Array rows).
function makeMelFilterbank(numBins, fftBins, sampleRate, lowFreq, highFreq) {
    const fftSize = (fftBins - 1) * 2;
    if (highFreq <= 0 || highFreq > sampleRate / 2) highFreq = sampleRate / 2;

    const melLow = hzToMel(lowFreq);
    const melHigh = hzToMel(highFreq);
    const melPoints = new Float64Array(numBins + 2);
    for (let i = 0; i < melPoints.length; i++) {
        melPoints[i] = melLow + (melHigh - melLow) * i / (numBins + 1);
    }
    const hzPoints = new Float64Array(melPoints.length);
    for (let i = 0; i < melPoints.length; i++) hzPoints[i] = melToHz(melPoints[i]);

    const binHz = new Float64Array(fftBins);
    for (let k = 0; k < fftBins; k++) binHz[k] = k * sampleRate / fftSize;

    const filt = [];
    for (let b = 0; b < numBins; b++) {
        const left = hzPoints[b], center = hzPoints[b + 1], right = hzPoints[b + 2];
        const row = new Float64Array(fftBins);
        for (let k = 0; k < fftBins; k++) {
            const f = binHz[k];
            if (f < left || f > right) continue;
            if (f <= center) row[k] = (f - left) / (center - left);
            else            row[k] = (right - f) / (right - center);
        }
        filt.push(row);
    }
    return filt;
}

// FbankComputer mirrors fbank.Fbank. Construct once, call compute() many.
// Not safe for concurrent use (scratch buffers).
export class FbankComputer {
    constructor(opts = fbankDefaults()) {
        this.opts = { ...fbankDefaults(), ...opts };
        const sr = this.opts.sampleRate;
        this.frameLength = Math.round(sr * this.opts.frameLengthMs / 1000);
        this.frameShift = Math.round(sr * this.opts.frameShiftMs / 1000);
        this.fftSize = nextPow2(this.frameLength);
        this.fftBins = (this.fftSize >> 1) + 1;
        const high = this.opts.highFreqHz > 0 ? this.opts.highFreqHz : sr / 2;
        this.window = makeWindow(this.frameLength, this.opts.windowType);
        this.mel = makeMelFilterbank(this.opts.numMelBins, this.fftBins, sr, this.opts.lowFreqHz, high);
        // Scratch.
        this._frameRe = new Float64Array(this.fftSize);
        this._frameIm = new Float64Array(this.fftSize);
        this._power = new Float64Array(this.fftBins);
    }

    numFrames(numSamples) {
        if (!this.opts.snipEdges) {
            return Math.floor((numSamples + this.frameShift / 2) / this.frameShift);
        }
        if (numSamples < this.frameLength) return 0;
        return Math.floor((numSamples - this.frameLength) / this.frameShift) + 1;
    }

    // compute returns a Float32Array of length numFrames * numMelBins, with
    // row-major layout: out[t*nbins + b]. The Go API returns [][]float32 but
    // the flat layout is friendlier for the LFR-stack + tensor copy that
    // follows in FSMN.
    compute(samples) {
        const nFrames = this.numFrames(samples.length);
        if (nFrames === 0) return new Float32Array(0);
        const nBins = this.opts.numMelBins;
        const out = new Float32Array(nFrames * nBins);
        for (let fi = 0; fi < nFrames; fi++) {
            const start = fi * this.frameShift;
            this._computeFrame(samples, start, out, fi * nBins);
        }
        return out;
    }

    _computeFrame(samples, start, out, outOff) {
        const re = this._frameRe;
        const im = this._frameIm;
        const N = this.fftSize;
        const L = this.frameLength;
        // Load frame, zero-pad tail.
        for (let i = 0; i < L; i++) re[i] = samples[start + i];
        for (let i = L; i < N; i++) re[i] = 0;
        for (let i = 0; i < N; i++) im[i] = 0;

        if (this.opts.removeDCOffset) {
            let mean = 0;
            for (let i = 0; i < L; i++) mean += re[i];
            mean /= L;
            for (let i = 0; i < L; i++) re[i] -= mean;
        }
        if (this.opts.preemphCoeff > 0) {
            const c = this.opts.preemphCoeff;
            // Kaldi: y[0] = x[0] - c*x[0]; y[i] = x[i] - c*x[i-1].
            let prev = re[0];
            re[0] = prev - c * prev;
            for (let i = 1; i < L; i++) {
                const cur = re[i];
                re[i] = cur - c * prev;
                prev = cur;
            }
        }
        for (let i = 0; i < L; i++) re[i] *= this.window[i];

        fftRadix2(re, im);

        // Power spectrum.
        const fftBins = this.fftBins;
        const power = this._power;
        for (let k = 0; k < fftBins; k++) power[k] = re[k] * re[k] + im[k] * im[k];

        // Mel * power + log. Matches kaldi-native-fbank: floor = float32-epsilon
        // before log, so silent frames log to ~-15.94.
        const FLOOR = 1.1920928955078125e-7;
        let floor = this.opts.energyFloor;
        if (floor < FLOOR) floor = FLOOR;
        const nBins = this.opts.numMelBins;
        for (let b = 0; b < nBins; b++) {
            const row = this.mel[b];
            let sum = 0;
            for (let k = 0; k < fftBins; k++) sum += row[k] * power[k];
            if (this.opts.useLog) {
                if (sum < floor) sum = floor;
                out[outOff + b] = Math.log(sum);
            } else {
                out[outOff + b] = sum;
            }
        }
    }
}

// lfrStack groups every `m` consecutive D-dim frames into one (m*D)-dim frame,
// sliding by `n` between outputs. If t < m, the last frame is replicated to
// reach length m (matches FunASR WavFrontend). Returns a Float32Array of
// shape [tLfr * m * d] row-major, plus tLfr.
//
// `feats` is a flat Float32Array of length t*d row-major.
export function lfrStack(feats, t, d, m, n) {
    if (t === 0) return { stacked: new Float32Array(0), tLfr: 0 };
    // Pad-by-replication if t < m. This means appending (m-t) copies of the
    // last d-vector at the end of `feats`.
    let frames = feats;
    if (t < m) {
        const padded = new Float32Array(m * d);
        padded.set(feats);
        const lastOff = (t - 1) * d;
        for (let i = t; i < m; i++) {
            for (let j = 0; j < d; j++) padded[i * d + j] = feats[lastOff + j];
        }
        frames = padded;
        t = m;
    }
    const tLfr = Math.floor((t - m) / n) + 1;
    const stacked = new Float32Array(tLfr * m * d);
    for (let i = 0; i < tLfr; i++) {
        for (let j = 0; j < m; j++) {
            const srcOff = (i * n + j) * d;
            const dstOff = i * (m * d) + j * d;
            for (let k = 0; k < d; k++) stacked[dstOff + k] = frames[srcOff + k];
        }
    }
    return { stacked, tLfr };
}

File: fbank/fbank.go
// Package fbank computes log Mel-filterbank features from 16 kHz mono audio,
// matching Kaldi's `compute-fbank-feats` defaults closely enough for VAD/ASR
// inputs. Parity against `kaldi_native_fbank` is verified in fbank_test.go.
//
// Used by the FSMN-VAD and FireRedVAD backends in pkg/vad. Speech models built
// on top of Kaldi-style features (which is most of them, including FunASR's
// FSMN-VAD and FireRedTeam's DFSMN-VAD) expect specific window shape, FFT
// size, Mel scale, and log floor; getting any of these wrong produces
// numerically-plausible-looking features that the model interprets as noise.
package fbank

import (
	"fmt"
	"math"
)

// Options configure a Fbank instance. Defaults() returns Kaldi's standard
// VAD/ASR settings; override individual fields as needed.
type Options struct {
	SampleRate     int
	FrameLengthMs  float64 // typical 25
	FrameShiftMs   float64 // typical 10
	NumMelBins     int     // typical 80
	WindowType     WindowType
	LowFreqHz      float64 // typical 20
	HighFreqHz     float64 // 0 means SampleRate/2
	PreemphCoeff   float64 // typical 0.97; set to 0 to disable
	RemoveDCOffset bool    // typical true
	SnipEdges      bool    // typical true; false = reflect-pad at edges
	EnergyFloor    float64 // typical 0; the natural-log floor applied to mel energies (clamped to >= this)
	UseLog         bool    // typical true; if false, return raw mel energies (rarely used)
}

// Defaults returns Kaldi compute-fbank-feats default options for 16 kHz audio
// with 80 mel bins and Povey window.
func Defaults() Options {
	return Options{
		SampleRate:     16000,
		FrameLengthMs:  25,
		FrameShiftMs:   10,
		NumMelBins:     80,
		WindowType:     WindowPovey,
		LowFreqHz:      20,
		HighFreqHz:     0, // => SampleRate/2
		PreemphCoeff:   0.97,
		RemoveDCOffset: true,
		SnipEdges:      true,
		EnergyFloor:    0,
		UseLog:         true,
	}
}

// Fbank computes Kaldi-compatible log-mel-filterbank features. Construct once
// (window + filterbank are precomputed); Compute is safe to call repeatedly.
// Not safe for concurrent calls on the same instance — clone if needed.
type Fbank struct {
	opts        Options
	frameLength int
	frameShift  int
	fftSize     int
	fftBins     int // fftSize/2 + 1
	window      []float64
	mel         [][]float64 // [NumMelBins][fftBins]

	// Scratch buffers reused across frames.
	frameBuf []float64
	fftBuf   []complex128
}

// New constructs a Fbank with the given options. Returns an error for
// degenerate options (e.g. zero frame length).
func New(opts Options) (*Fbank, error) {
	if opts.SampleRate <= 0 {
		return nil, fmt.Errorf("fbank: SampleRate must be > 0")
	}
	if opts.FrameLengthMs <= 0 || opts.FrameShiftMs <= 0 {
		return nil, fmt.Errorf("fbank: FrameLengthMs and FrameShiftMs must be > 0")
	}
	if opts.NumMelBins <= 0 {
		return nil, fmt.Errorf("fbank: NumMelBins must be > 0")
	}
	frameLength := int(math.Round(float64(opts.SampleRate) * opts.FrameLengthMs / 1000))
	frameShift := int(math.Round(float64(opts.SampleRate) * opts.FrameShiftMs / 1000))
	fftSize := nextPow2(frameLength)
	fftBins := fftSize/2 + 1

	high := opts.HighFreqHz
	if high <= 0 {
		high = float64(opts.SampleRate) / 2
	}

	return &Fbank{
		opts:        opts,
		frameLength: frameLength,
		frameShift:  frameShift,
		fftSize:     fftSize,
		fftBins:     fftBins,
		window:      makeWindow(frameLength, opts.WindowType),
		mel:         makeMelFilterbank(opts.NumMelBins, fftBins, float64(opts.SampleRate), opts.LowFreqHz, high),
		frameBuf:    make([]float64, fftSize),
		fftBuf:      make([]complex128, fftSize),
	}, nil
}

// NumFrames returns how many output frames Compute will produce for a sample
// vector of the given length.
func (f *Fbank) NumFrames(numSamples int) int {
	if f.opts.SnipEdges {
		if numSamples < f.frameLength {
			return 0
		}
		return (numSamples-f.frameLength)/f.frameShift + 1
	}
	// snip_edges=false: Kaldi uses center-frame alignment with reflection padding.
	// Not used by our current backends; not implemented here.
	return (numSamples + f.frameShift/2) / f.frameShift
}

// Compute returns [NumFrames(len(samples))][NumMelBins] log-mel features.
// The input is mono PCM, float32 valued in the same scale as Kaldi expects
// (e.g. raw int16 values cast to float, or float32 PCM multiplied by 32768
// to match — FunASR and FireRed front-ends both scale [-1,1] floats up).
func (f *Fbank) Compute(samples []float32) [][]float32 {
	nFrames := f.NumFrames(len(samples))
	if nFrames == 0 {
		return nil
	}
	out := make([][]float32, nFrames)
	for i := range out {
		out[i] = make([]float32, f.opts.NumMelBins)
	}

	for fi := 0; fi < nFrames; fi++ {
		start := fi * f.frameShift
		f.computeFrame(samples[start:start+f.frameLength], out[fi])
	}
	return out
}

func (f *Fbank) computeFrame(in []float32, out []float32) {
	// Copy + remove DC + preemphasize, then window.
	for i := 0; i < f.frameLength; i++ {
		f.frameBuf[i] = float64(in[i])
	}
	for i := f.frameLength; i < f.fftSize; i++ {
		f.frameBuf[i] = 0
	}

	if f.opts.RemoveDCOffset {
		var mean float64
		for i := 0; i < f.frameLength; i++ {
			mean += f.frameBuf[i]
		}
		mean /= float64(f.frameLength)
		for i := 0; i < f.frameLength; i++ {
			f.frameBuf[i] -= mean
		}
	}

	if f.opts.PreemphCoeff > 0 {
		// Kaldi pre-emphasis: y[0] = x[0] - coeff*x[0]; y[i] = x[i] - coeff*x[i-1]
		coeff := f.opts.PreemphCoeff
		prev := f.frameBuf[0]
		f.frameBuf[0] = prev - coeff*prev
		for i := 1; i < f.frameLength; i++ {
			cur := f.frameBuf[i]
			f.frameBuf[i] = cur - coeff*prev
			prev = cur
		}
	}

	for i := 0; i < f.frameLength; i++ {
		f.frameBuf[i] *= f.window[i]
	}

	// Real → complex FFT input.
	for i := 0; i < f.fftSize; i++ {
		f.fftBuf[i] = complex(f.frameBuf[i], 0)
	}
	fftRadix2(f.fftBuf)

	// Power spectrum: |X[k]|^2 for the first fftBins (positive-freq half).
	power := make([]float64, f.fftBins) // small alloc; we could pool if hot
	for k := 0; k < f.fftBins; k++ {
		re := real(f.fftBuf[k])
		im := imag(f.fftBuf[k])
		power[k] = re*re + im*im
	}

	// Mel filterbank * power, then log. Match kaldi-native-fbank, which uses
	// std::numeric_limits<float>::epsilon() (≈ 1.1920929e-7) as the implicit
	// floor before log — so log of "silent" frames is -15.942 (not -23 or
	// whatever a smaller floor would produce). Tested via parity fixture.
	const float32Epsilon = 1.1920928955078125e-7
	floor := f.opts.EnergyFloor
	if floor < float32Epsilon {
		floor = float32Epsilon
	}
	for b := 0; b < f.opts.NumMelBins; b++ {
		row := f.mel[b]
		var sum float64
		for k := 0; k < f.fftBins; k++ {
			sum += row[k] * power[k]
		}
		if f.opts.UseLog {
			if sum < floor {
				sum = floor
			}
			out[b] = float32(math.Log(sum))
		} else {
			out[b] = float32(sum)
		}
	}
}
